core/src/test/java/org/apache/iceberg/TestBaseIncrementalChangelogScan.java

```diff
@@ -132,6 +131,175 @@ public void testFileDeletes() {
     assertThat(t1.existingDeletes()).as("Must be no deletes").isEmpty();
   }
 
+  @TestTemplate
+  public void testRowDeletes() {
+    assumeThat(formatVersion).isEqualTo(2);
+
+    table
+        .newFastAppend()
+        .appendFile(FILE_A)
+        .appendFile(FILE_A2)
+        .appendFile(FILE_B)
+        .appendFile(FILE_C)
+        .commit();
+    Snapshot snap1 = table.currentSnapshot();
+
+    // position delete
+    table.newRowDelta().addDeletes(FILE_B_DELETES).commit();
+    Snapshot snap2 = table.currentSnapshot();
+
+    // equality delete
+    table.newRowDelta().addDeletes(FILE_C2_DELETES).commit();
+    Snapshot snap3 = table.currentSnapshot();
+
+    // mix of position and equality deletes
+    table.newRowDelta().addDeletes(FILE_A_DELETES).addDeletes(FILE_A2_DELETES).commit();
```

Comment:
One more interesting question is:
- How we will handle the compaction commits? Is there a better way than emit every insert/delete record? Could we skip it?

Is it trivial to fix in this PR, or is that need another PR?

Reply:
I'm not sure I understand your question. Are you referring to outputting *net changes*? If so, I think that should be a separate PR. I think for that, we should introduce an option for it.
If you're referring to something else, can you please explain?

Reply:
ps. Net changes is something a user might want regardless of whether merge-on-read or copy-on-write is used, so is an orthogonal concern.

Reply:
This is the question about how you see the compaction where we rewrite the files with the expressed goal to keep the actual content unchanged.
- Is this a change where we remove and add files which are coincidentally contain the same data?
- Is this a change where no data has changed?

I see the reasoning behind both cases, but using the first approach would most probably make the changelog scan unfesable for cases where compaction happens on the table.

Reply:
Ah, you mean compaction as in rewriting data files.
In this case, compaction does not result in any inserts or deletes. (I believe this is what we both expect.) Please see `testDataRewritesAreIgnored` that I added in `TestChangelogTable` in this PR. This test is in spark-extensions because I want to call the rewrite_data_files Spark procedure.

Reply:
Cool!
Thanks for the explanation!